platforms/core-configuration/model-core/src/testFixtures/groovy/org/gradle/api/internal/provider/AbstractPropertySpec.groovy

```diff
@@ -124,4 +126,145 @@ abstract class AbstractPropertySpec<T> extends PropertySpec<T> {
+
+        expect:
+        property.type == property.shallowCopy().type
+    }
+
+    def "shallow copy inherits dependencies of the original"() {
+        given:
+        def task = Mock(Task)
+        def provider = supplierWithProducer(task)
+        def property = propertyWithNoValue()
+        property.set(provider)
+
+        expect:
+        assertHasProducer(property.shallowCopy(), task)
+    }
+
+    def "shallow copy does not follow changes to dependencies of the original"() {
+        given:
+        def task = Mock(Task)
+        def provider = supplierWithProducer(task)
+        def property = propertyWithNoValue()
+        property.set(provider)
+
+        when:
+        def copy = property.shallowCopy()
+        property.set(someOtherValue())
+
+        then:
+        assertHasProducer(copy, task)
+        assertHasNoProducer(property)
+    }
+
+    def "shallow copy does not inherit producer from the original"() {
+        given:
+        def task = Mock(Task)
+        def property = propertyWithNoValue()
+        property.attachProducer(owner(task))
+
+        expect:
+        assertHasNoProducer(property.shallowCopy())
+    }
+
+    def "shallow copy reflects changes to the value"() {
+        given:
+        def innerProperty = propertyWithValue(someValue())
+        def property = propertyWithNoValue()
+        property.set(innerProperty)
+
+        when:
+        def copy = property.shallowCopy()
+        innerProperty.set(someOtherValue())
+
+        then:
+        copy.orNull == innerProperty.orNull
+    }
+
+    def "shallow copy is fixed if the underlying provider is fixed"() {
+        given:
+        def property = propertyWithValue(someValue())
+
+        when:
+        def copy = property.shallowCopy()
+
+        then:
+        !copy.calculateExecutionTimeValue().isChangingValue()
+    }
+
+    def "shallow copy is changing if the underlying provider is changing"() {
+        given:
+        def upstream = supplierWithChangingExecutionTimeValues(someValue())
+        def property = propertyWithNoValue()
+        property.set(upstream)
+
+        when:
+        def copy = property.shallowCopy()
+
+        then:
+        copy.calculateExecutionTimeValue().isChangingValue()
+    }
+
+    def "shallow copy copies the convention if the property has no explicit value"() {
+        given:
+        def property = propertyWithNoValue()
+        property.convention(someValue())
+
+        expect:
+        property.shallowCopy().orNull == someValue()
+    }
+
+    def "shallow copy is not affected by changes to original convention"() {
+        given:
+        def property = propertyWithNoValue()
+        property.convention(someValue())
+
+        when:
+        def copy = property.shallowCopy()
+        property.convention(someOtherValue())
+
+        then:
+        copy.orNull == someValue()
+    }
+
+    def "obtaining value of the shadow copy does not finalize property"() {
```

Comment:
👍 

This also means that a provider given to a `finalizeValueOnRead` property could end up being executed multiple times via a shallow copy, right?

Reply:
Indeed

Reply:
Because the shallow copy is a valueState instance that is independent of the base property value:

https://github.com/gradle/gradle/blob/0bb7c093f832ca638250e9aa395781b61cbd4603/platforms/core-configuration/model-core/src/main/java/org/gradle/api/internal/provider/AbstractProperty.java#L336